src/dbnode/storage/index/compaction/plan.go

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package compaction
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/storage/index/segments"
+)
+
+var (
+	errMutableCompactionAgeNegative = errors.New("mutable compaction age must be postive")
+	errSizeBucketsUndefined         = errors.New("size buckets are undefined")
+	errMaxImmutableCompactionSize   = errors.New("max immutable compaction size must be positive")
+)
+
+var (
+	DefaultSizeBuckets = []SizeBucket{ // i.e. tiers for compaction [0, 262K), [262K, 524K), [524K, 1M), [1M, 2M), [2M, 8M)
+		SizeBucket{
+			MinSizeInclusive: 0,
+			MaxSizeExclusive: 1 << 18,
+		},
+		SizeBucket{
+			MinSizeInclusive: 1 << 18,
+			MaxSizeExclusive: 1 << 19,
+		},
+		SizeBucket{
+			MinSizeInclusive: 1 << 19,
+			MaxSizeExclusive: 1 << 20,
+		},
+		SizeBucket{
+			MinSizeInclusive: 1 << 20,
+			MaxSizeExclusive: 1 << 21,
+		},
+		SizeBucket{
+			MinSizeInclusive: 1 << 21,
+			MaxSizeExclusive: 1 << 23,
+		},
+	}
+
+	DefaultOptions = PlannerOptions{
+		MaxImmutableCompactionSize: 1 << 22,                            // ~4M
+		MaxMutableSegmentSize:      1 << 16,                            // 64K
+		MutableCompactionAge:       15 * time.Second,                   // any mutable segment 15s or older is eligible for compactions
```

Comment:
I wonder if in the future we can just get rid of this option and if we ever have a mutable segment, always start compacting.

Reply:
Would setting an age of 0 here make this work like so?

Reply:
I tried the equivalent - creating a mutable segment every time a write batch came in, and triggering a compaction - it ended up creating a lot small segments and the test nodes compactions couldn't keep up. A buffer of even 5s helped this a lot. 

Is the concern e2e latency?